File: Anul 2/Semestrul 2/MPP/labs/Tema 5-proiect Server Csharp-Client Java/proiectSCcsharp/proiectServerClientCsharp/Service.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static System.Windows.Forms.VisualStyles.VisualStyleElement.StartPanel;
using System.Xml.Linq;
using System.Data.Entity.Core.Objects;
using proiectSCcsharp.domain.validators;
using proiectSCcsharp.domain;
using proiectSCcsharp.repository;
using proiectSCcsharp.services;

namespace proiectSCcsharp.server
{
    public class Service : IService
    {
        private readonly Validator<Angajat> angajatValidator;
        private readonly Validator<Spectacol> spectacolValidator;
        private readonly Validator<Bilet> biletValidator;
        private readonly RepoAngajat angajatRepo;
        private readonly RepoSpectacol spectacolRepo;
        private readonly RepoBilet biletRepo;
        private readonly IDictionary<long, IObserver> loggedAngajati;

        public Service(Validator<Angajat> angajatValidator, Validator<Spectacol> spectacolValidator, Validator<Bilet> biletValidator, RepoAngajat repoAngajat, RepoSpectacol repoSpectacol, RepoBilet repoBilet)
        {
            this.angajatValidator = angajatValidator;
            this.biletValidator = biletValidator;
            this.spectacolValidator = spectacolValidator;
            this.angajatRepo = repoAngajat;
            this.spectacolRepo = repoSpectacol;
            this.biletRepo = repoBilet;
            loggedAngajati = new Dictionary<long, IObserver>();
        }

        public Spectacol getSpectacol(long id)
        {
            return spectacolRepo.findElement(id);
        }
        
        public List<Spectacol> getAllSpectacole()
        {
            return spectacolRepo.getAll();
        }
        
        public void addBilet(Bilet bilet)
        {
            if (getSpectacol(bilet.idSpectacol).availableSeats>=bilet.nrSeats)
            {
                Spectacol spectacol = getSpectacol(bilet.idSpectacol);
                biletRepo.add(bilet);
                spectacolRepo.update(spectacol, bilet.nrSeats);
                foreach (var employeeO in loggedAngajati.Values)
                {
                    Task.Run(() => employeeO.BiletAdded(bilet));
                }
            }
            else
            {
                throw new MyException("Nu exista atatea locuri disponibile.");
            }
        }

        public Angajat login(Angajat employee, IObserver employeeObserver)
        {
            var authenticateAngajat = angajatRepo.AuthenticateAngajat(employee.username, employee.password);
            if (authenticateAngajat != null)
            {
                if (loggedAngajati.ContainsKey(authenticateAngajat.id))
                    throw new MyException("User already logged in.");
                loggedAngajati[authenticateAngajat.id] = employeeObserver;
            }
            else
                throw new MyException("Authentication failed.");
            return authenticateAngajat;
        }

        public void logout(Angajat employee)
        {
            var localClient = loggedAngajati[employee.id];
            if (localClient == null)
                throw new MyException("User " + employee.id + " is not logged in.");
            loggedAngajati.Remove(employee.id);
        }
    }
}

File: Anul 2/Semestrul 2/MPP/labs/Tema 7-servicii REST/Java/Rest/proiect/Server/src/main/java/proiectServerClientJava/server/Service.java
package proiectServerClientJava.server;

import proiect.domain.Angajat;
import proiect.domain.Bilet;
import proiect.domain.Spectacol;
import proiect.domain.validators.ValidationException;
import proiect.domain.validators.Validator;
import proiect.persistence.repository.jdbc.RepoAngajat;
import proiect.persistence.repository.jdbc.RepoBilet;
import proiect.persistence.repository.jdbc.RepoSpectacol;
import proiect.service.IObserver;
import proiect.service.IService;
import proiect.service.MyException;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class Service implements IService{
    private final Validator<Angajat> angajatValidator;
    private final Validator<Spectacol> spectacolValidator;
    private final Validator<Bilet> biletValidator;
    private final RepoAngajat angajatRepo;
    private final RepoSpectacol spectacolRepo;
    private final RepoBilet biletRepo;
    private Map<Integer, IObserver> loggedAngajati;


    public Service(Validator<Angajat> validatorA,
                   Validator<Spectacol> validatorS,
                   Validator<Bilet> validatorB,
                   RepoAngajat angajatRepo,
                   RepoSpectacol spectacolRepo,
                   RepoBilet biletRepo)
    {
        this.angajatValidator=validatorA;
        this.spectacolValidator=validatorS;
        this.biletValidator=validatorB;
        this.angajatRepo=angajatRepo;
        this.spectacolRepo=spectacolRepo;
        this.biletRepo=biletRepo;
        loggedAngajati =new ConcurrentHashMap<>();
    }

    @Override
    public Spectacol getSpectacol(int id){
        return spectacolRepo.findElement(id);
    }

    @Override
    public List<Spectacol> getAllSpectacole() {
        return spectacolRepo.getAll();
    }

    @Override
    public void addBilet(Bilet bilet) throws MyException, ValidationException {
        if(getSpectacol(bilet.getIdSpectacol()).getAvailableSeats()>=bilet.getNrSeats())
        {
            Spectacol spectacol=getSpectacol(bilet.getIdSpectacol());
            biletValidator.validate(bilet);
            biletRepo.add(bilet);
            spectacolRepo.update(spectacol,bilet.getNrSeats());
            for (IObserver loggedUser: loggedAngajati.values()) {
                loggedUser.biletAdded(bilet);
            }
        }
        else
        {
         throw new MyException("Nu exista atatea locuri disponibile.");
        }
    }

    public synchronized Angajat login(Angajat employee, IObserver client) throws Exception {
        Angajat employeeToLogin=angajatRepo.authenticateAngajat(employee.getUsername(),employee.getPassword());
        if (employeeToLogin!=null){
            if (loggedAngajati.get(employeeToLogin.getId())!=null)
                throw new Exception("Angajat already logged in.");
            loggedAngajati.put(employeeToLogin.getId(),client);
        }
        else
            throw new Exception("Authentication failed.");
        return employeeToLogin;
    }
    public synchronized void logout(Angajat employee, IObserver client) throws Exception {
        IObserver loggedEmployee= loggedAngajati.remove(employee.getId());
        if (loggedEmployee==null)
            throw new Exception("Employee "+String.valueOf(employee.getId())+" is not logged in.");
    }
}
